### ISSUE TITLE
undefined symbol: std::__1::__libcpp_verbose_abort(char const*, ...) in debug builds of libc++abi.so

### ISSUE THREAD

**Issue by arichardson** [llvm/llvm-project#58117]
When trying to build the runtime libraries with `-DCMAKE_BUILD_TYPE=Debug`, I get linker errors in libc++abi.so:

Build command: `mkdir test-build && cmake -S ./runtimes -B ./test-build -GNinja -DCMAKE_BUILD_TYPE=Debug -DLIBCXX_ENABLE_ASSERTIONS=TRUE "-DLLVM_ENABLE_RUNTIMES=libunwind;libcxxabi;libcxx" -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_C_COMPILER=/usr/bin/clang && ninja -C test-build`

```
libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_demangle.cpp.o: in function `void std::__1::advance[abi:v160000]<(anonymous namespace)::itanium_demangle::AbstractManglingParser<(anonymous namespace)::itanium_demangle::ManglingParser<(anonymous namespace)::DefaultAllocator>, (anonymous namespace)::DefaultAllocator>::OperatorInfo const*, long, long, void>((anonymous namespace)::itanium_demangle::AbstractManglingParser<(anonymous namespace)::itanium_demangle::ManglingParser<(anonymous namespace)::DefaultAllocator>, (anonymous namespace)::DefaultAllocator>::OperatorInfo const*&, long)':
cxa_demangle.cpp:(.text+0x195e0): undefined reference to `std::__1::__libcpp_verbose_abort(char const*, ...)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by arichardson** [comment]
It appears this call comes from the call to std::lower_bound() inside `AbstractManglingParser<Derived, Alloc>::parseOperatorEncoding()`, which was added in 7f89fa32e8e9251b0b780009c340bc0d8f4d2e88. With the call to lower_bound commented out, I can successfully build the library.

I guess this means CI coverage for `-DCMAKE_BUILD_TYPE=Debug -DLIBCXX_ENABLE_ASSERTIONS=TRUE` is missing?

CC: @urnathan

**Comment by arichardson** [comment]
I can see multiple solutions for this issue, but I'm not sure what would be best:
- we could add #define _LIBCPP_ENABLE_ASSERTIONS 0` at the top of cxa_demangle.cpp (uploaded as https://reviews.llvm.org/D135068). However, that means disabling potentially useful assertions.
- We could add an inline implementation of std::__1::__libcpp_verbose_abort in cxa_demangle.cpp. Keeps assertions but adds extra code
- We could re-implement std::lower_bound in cxa_demangle.cpp without calls to libc++ assertions.

Not sure what would be the best solution here.

**Comment by urnathan** [comment]
ISTM that a debug libcxx-asserting build should provide the library containing its asserting machinery to the shared object link of the demangler.  Why would it not do so?
It seems really awkward to forbid a (undocumented?) subset of the STL from the demangler's implementation.

**Comment by arichardson** [comment]
I agree this is awkward. It appears the restriction is: You can use all of libc++ as long as it doesn't result in a call to a function defined in the shared library. From demangler/README.txt:

> Because the core demangler needs to work in libcxxabi, everything
> needs to be declared in an anonymous namespace (see
> DEMANGLE_NAMESPACE_BEGIN), and you can't introduce any code that
> depends on the libcxx dylib.

**Comment by arichardson** [comment]
> d'oh! ok, let's open code binary search. That's very straight forwards. do you want to do that, or assign this to me?

I tried the simple approach and that appears to break CI (https://reviews.llvm.org/D135068), and unfortunately don't have time to continue working on this (at least for this week), so if you could do it that would be great :)
